openassessment/xblock/apis/submissions/submissions_actions.py

```diff
@@ -256,3 +261,97 @@ def save_submission_draft(student_submission_data, block_config_data, block_subm
         )
     except Exception as e:  # pylint: disable=broad-except
         raise DraftSaveException from e
+
+
+def append_file_data(file_data, block_config, submission_info):
+    """
+    Appends a list of file data to the current block state
+
+    Args:
+        block_config (ORAConfigAPI)
+        submission_info (SubmissionAPI)
+        files_to_append (list of {
+            'description': (str)
+            'name': (str)
+            'size': (int)
+        })
+    """
+    try:
+        submission_info.files.file_manager.append_uploads(*file_data)
+        # Emit analytics event...
+        block_config.publish_event(
+            "openassessmentblock.save_files_descriptions",
+            {"saved_response": submission_info.files.saved_files_descriptions},
+        )
+    except FileUploadError as exc:
+        logger.exception(
+            "append_file_data: file description for data %s failed with error %s", file_data, exc, exc_info=True
+        )
+        raise
+    except Exception as exc:  # pylint: disable=broad-except
+        logger.exception(
+            "append_file_data: unhandled exception for data %s. Error: %s", file_data, exc, exc_info=True
+        )
+        raise FileUploadError(exc) from exc
+
+
+def remove_uploaded_file(file_index, block_config, submission_info):
+    """
+    Removes uploaded user file.
+    """
+    file_key = submission_info.files.get_file_key(file_index)
+    if not submission_info.files.can_delete_file(file_index):
+        raise DeleteNotAllowed()
+    try:
+        submission_info.files.delete_uploaded_file(file_index)
+        # Emit analytics event...
+        block_config.publish_event(
+            "openassessmentblock.remove_uploaded_file",
+            {"student_item_key": file_key},
+        )
+        logger.debug("Deleted file %s", file_key)
+    except FileUploadError as exc:
+        logger.exception(
+            "FileUploadError: Error when deleting file %s : %s",
+            file_key,
+            exc,
+            exc_info=True,
+        )
+        raise
+    except Exception as exc:  # pylint: disable=broad-except
+        logger.exception(
+            "FileUploadError: unhandled exception for %s. Error: %s",
+            file_key,
+            exc,
+            exc_info=True,
+        )
+        raise FileUploadError(exc) from exc
+
+
+def get_upload_url(content_type, file_name, file_index, block_config, submission_info):
+    """
+    Request a URL to be used for uploading content for a given file
+
+    Returns:
+        A URL to be used to upload content associated with this submission.
+
+    """
+    if not block_config.allow_multiple_files:
+        if submission_info.files.has_any_file_in_upload_space():
```

Comment:
Why is this checking with `not allow_multiple_files` then check max file in the function? What if the `allow_multiple_files = 2`? Unless that isn't an option.

Reply:
allow_multiple_files is just a boolean, it's only gonna be false (only one file) or true (as many as we allow)
we don't current;y allow control of # of submissions beyond that

Reply:
this is a really janky use case that gets really into the under the hood implementation details of file uploads. Rather than being able to trust the block-level fields, we basically hit every possible file key that could exist for this submission on the file backend and see if they exist

Reply:
Sorry, I was trying to say `max_allowed_uploads`.